runtime/vm/ObjectMonitor.cpp

```diff
@@ -406,9 +521,40 @@ spinOnFlatLock(J9VMThread *currentThread, j9objectmonitor_t volatile *lwEA, j9ob
 				goto done;
 			}
 			/* do not spin if the FLC, inflated or reserved bits are already set */
-			if (J9_ARE_NO_BITS_SET(J9_LOAD_LOCKWORD(currentThread, lwEA), bits)
-					&& J9_ARE_NO_BITS_SET(currentThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_EXCLUSIVE))
-			{
+			j9objectmonitor_t const lock = J9_LOAD_LOCKWORD(currentThread, lwEA);
+			if (J9_ARE_NO_BITS_SET(lock, bits) && J9_ARE_NO_BITS_SET(currentThread->publicFlags, J9_PUBLIC_FLAGS_HALT_THREAD_EXCLUSIVE)) {
+				/* If the Learning bit is set, need to handle Learning state. */
+				if (0 != (lock & OBJECT_HEADER_LOCK_LEARNING)) {
+					/* Check if RC is 0, if so it is possible to just atomically lock the object now. */
+					if (0 == J9_FLATLOCK_COUNT(lock)) {
+						if (lock == VM_ObjectMonitor::compareAndSwapLockword(currentThread, lwEA, lock, (j9objectmonitor_t)(UDATA)currentThread, false)) {
+							/* compare and swap succeeded */
+							VM_AtomicSupport::monitorEnterBarrier();
+							rc = true;
+
+							/* Transition from Learning to Flat occurred so the Cancel Counter in the object's J9Class is incremented by 1. */
+							VM_ObjectMonitor::incrementCancelCounter(J9OBJECT_CLAZZ(currentThread, object));
+							goto done;
+						}
+					} else {
+						/*
+						 * Attempt to atomically change the lock for Learning to Flat. Try again next iteration if this doesn't work.
+						 * Clears the Learning bit to convert the lock to Flat state.
+						 * Clears the bits under OBJECT_HEADER_LOCK_RECURSION_MASK then fills in the new RC.
+						 * This is done to reclaim the bits used for the learning counter that are no longer used under Flat.
+						 * For the same logical recursion count, the value encoded in the RC bits is one lower under
+						 * Flat compared to Learning so the new RC field is one lower than before the conversion to flat.
+						 */
+						j9objectmonitor_t const newLock = (lock & ~((j9objectmonitor_t)OBJECT_HEADER_LOCK_LEARNING | (j9objectmonitor_t)OBJECT_HEADER_LOCK_RECURSION_MASK)) |
```

Comment:
Merge casts:
```
j9objectmonitor_t const newLock = (lock & ~(j9objectmonitor_t)(OBJECT_HEADER_LOCK_LEARNING | OBJECT_HEADER_LOCK_RECURSION_MASK)) |

```

Reply:
I changed that. I also found a few other places where the casts could be combined and also combined them.